### PR TITLE
Enable joins for hierarchical continuous aggregates

### PR DESCRIPTION
**Pull request by RafiaSabih**
The joins could be between a continuous aggregate and hypertable, continuous aggregate and a regular Postgres table, and continuous aggregate and a regular Postgres view.

### CODE REVIEW

File: tsl/src/continuous_aggs/create.c [hunk 1]
```diff
@@ -1287,10 +1287,10 @@ cagg_validate_query(const Query *query, const bool finalized, const char *cagg_s
...
 		 * Cagg with joins does not support hierarchical caggs in from clause.
 		 */
-		if (rte->relkind == RELKIND_VIEW || rte_other->relkind == RELKIND_VIEW)
-			ereport(ERROR,
-					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					 errmsg("joins for hierarchical continuous aggregates are not supported")));
+		// if (rte->relkind == RELKIND_VIEW || rte_other->relkind == RELKIND_VIEW)
+		// 	ereport(ERROR,
+		// 			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+		// 			 errmsg("joins for hierarchical continuous aggregates are not supported")));
```
Comment (fabriziomello) [inline tsl/src/continuous_aggs/create.c:1293]: Leftover?

File: CHANGELOG.md [hunk 2]
```diff
@@ -43,6 +43,7 @@ Sooner to that time, we will announce the specific version of TimescaleDB in whi
 * #5343 Set PortalContext when starting job
 * #5312 Add timeout support to the ping_data_node()
 * #5212 Allow pushdown of reference table joins 
+* #5344 Enable joins for hierarchical continuous aggregates
```
Comment (fabriziomello) [inline CHANGELOG.md:46]:
```suggestion
* #5344 Enable JOINS for Hierarchical Continuous Aggregates
```

File: tsl/test/expected/cagg_joins-12.out [hunk 3]
```diff
@@ -310,7 +310,8 @@ SELECT time_bucket(INTERVAL '1 day', day) AS bucket,
 FROM conditions, devices_view
 WHERE conditions.device_id = devices_view.device_id
 GROUP BY name, bucket, devices_view.device_id;
-ERROR:  joins for hierarchical continuous aggregates are not supported
+NOTICE:  refreshing continuous aggregate "conditions_summary_daily_view"
```
Comment (kgyrtkirk) [inline tsl/test/expected/cagg_joins-12.out:313]: I see that these statements need the mentioned feature - but don't we need real queries/plan-s to track that it works as expected?

Comment (fabriziomello) [reply tsl/test/expected/cagg_joins-12.out:313]: I agree with @kgyrtkirk and wonder if we should add some JOIN tests in the caggs_on_caggs suite where we extensively test it checking plans, doing different time bucket data type tests... caggs on caggs validation tests... etc. WDYT @RafiaSabih ?

File: CHANGELOG.md [hunk 4]
```diff
@@ -63,6 +63,10 @@ Sooner to that time, we will announce the specific version of TimescaleDB in whi
 * #5246 Make connection establishment interruptible
 * #5253 Make data node command execution interruptible
 * #5262 Extend enabling compression on a continuous aggregrate with 'compress_segmentby' and 'compress_orderby' parameters
+* #5343 Set PortalContext when starting job
+* #5312 Add timeout support to the ping_data_node()
+* #5212 Allow pushdown of reference table joins
+* #5344 Enable JOINS for Hierarchical continuous aggregates
```
Comment (fabriziomello) [inline CHANGELOG.md:69]: I guess your CHANGELOG entry is on the wrong place, it should reside on the *Unreleased* section at the beginning of the file.

**Review by kgyrtkirk — APPROVED** [review]

File: tsl/test/sql/include/cagg_on_cagg_setup.sql [hunk 5]
```diff
@@ -13,15 +13,37 @@ SET ROLE :ROLE_DEFAULT_PERM_USER;
...
     temperature NUMERIC
   );
+  \endif
 \else
-  CREATE TABLE conditions (
+  \if :IS_JOIN
+    CREATE TABLE conditions (
+    temperature NUMERIC,
+    time :TIME_DIMENSION_DATATYPE NOT NULL,
+    device_id INT
```
[the remark this answers is not in the document]
Comment (fabriziomello) [reply tsl/test/sql/include/cagg_on_cagg_setup.sql:33]: Good point @kgyrtkirk ... don't need to add those extra conditionals for new column, just add it to the original schema.

File: tsl/test/sql/include/cagg_on_cagg_common.sql [hunk 6]
```diff
@@ -6,31 +6,74 @@
 CREATE MATERIALIZED VIEW :CAGG_NAME_1ST_LEVEL
 WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS
 SELECT
+\if :IS_JOIN
+  time_bucket(:BUCKET_WIDTH_1ST, "time") AS bucket,
+  SUM(temperature) AS temperature,
+  device_id
+\else
```
Comment (fabriziomello) [inline tsl/test/sql/include/cagg_on_cagg_common.sql:13]: Insted of all this conditionals what about something like this in the tests that include `cagg_on_cagg_common.sql`
```sql
-- In the cagg_on_cagg.sql
\set EXTRA_GROUPBY_COLUMN
-- In the cagg_on_cagg_joins.sql
\set EXTRA_GROUPBY_COLUMN ,device_id
```

And then use it in the query in `cagg_on_cagg_common.sql`:
```suggestion
  time_bucket(:BUCKET_WIDTH_1ST, "time") AS bucket,
  SUM(temperature) AS temperature
  :EXTRA_GROUPBY_COLUMN
```
Doing that you'll reduce the log output.

**Comment by fabriziomello** [comment]
@RafiaSabih you also should add tests for distributed hypertables like we do in `cagg_on_cagg_dist_ht.sql`

File: tsl/test/sql/include/cagg_on_cagg_setup.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (RafiaSabih) [reply tsl/test/sql/include/cagg_on_cagg_setup.sql:25]: Done that now. Thanks for the suggestion. I was trying to limit the changes hence did that, but this way it is more readable, so better.

File: tsl/src/continuous_aggs/create.c [hunk 7]
```diff
@@ -1283,31 +1284,34 @@ cagg_validate_query(const Query *query, const bool finalized, const char *cagg_s
...
+			(rte->relkind == RELKIND_VIEW && ts_is_hypertable(rte_other->relid)) ||
+			(rte_other->relkind == RELKIND_VIEW && ts_is_hypertable(rte->relid)) ||
+			(rte->relkind != RELKIND_VIEW && rte_other->relkind != RELKIND_VIEW &&
+			 (ts_is_hypertable(rte->relid) == ts_is_hypertable(rte_other->relid))))
 			ereport(ERROR,
 					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 					 errmsg("invalid continuous aggregate view"),
-					 errdetail("from clause can only have one hypertable and one normal table")));
+					 errdetail("views, lateral, multiple hypertables or normal tables"
+							   " are not supported in FROM clause")));
```
Comment (fabriziomello) [inline tsl/src/continuous_aggs/create.c:1314]: This branch is a bit hard to follow... isn't better to split it in separated checks and with specific message and test case for each of them??

Comment (RafiaSabih) [reply tsl/src/continuous_aggs/create.c]: Thanks for your review, I have broken it in multiple conditions. Regarding the test, they are already covered.

**Comment by konskov** [comment]
looks like there is a spelling error in the commit message (heirarchical instead of hierarchical)?
